clippy_utils/src/ast_utils.rs

```diff
@@ -571,3 +572,36 @@ pub fn eq_mac_args(l: &MacArgs, r: &MacArgs) -> bool {
         _ => false,
     }
 }
+
+/// Extract args from an assert-like macro.
+/// Currently working with:
+/// - `assert!`, `assert_eq!` and `assert_ne!`
+/// - `debug_assert!`, `debug_assert_eq!` and `debug_assert_ne!`
+/// For example:
+/// `assert!(expr)` will return Some([expr])
+/// `debug_assert_eq!(a, b)` will return Some([a, b])
+pub fn extract_assert_macro_args(e: &Expr) -> Option<Vec<&P<Expr>>> {
+    fn get_args(e: &Expr) -> Option<Vec<&P<Expr>>> {
+        if_chain! {
+            if let ExprKind::Block(ref block, _) = e.kind;
+            if let StmtKind::Expr(ref expr) = block.stmts.get(0)?.kind;
+            if let ExprKind::Match(ref match_expr, _) = expr.kind;
+            if let ExprKind::Tup(ref tup) = match_expr.kind;
+            then {
+                return Some(tup.iter().filter_map(|t| match t.kind {
+                    ExprKind::AddrOf(_, _, ref expr) => Some(expr),
+                    _ => None,
+                }).collect::<Vec<_>>());
+            }
+        }
+        None
+    }
+    if_chain! {
+        if let ExprKind::If(_, ref block, _) = e.kind;
+        if let StmtKind::Semi(ref e) = block.stmts.get(0)?.kind;
+        then {
+            return get_args(e);
```

Comment:
You can easily avoid having an inner function like this.

```suggestion
            expr = e;
```

Reply:
Good idea!